docs/rfc/rfc-vpf-001-question-routing.md

```diff
@@ -0,0 +1,88 @@
+
+```ruby
+  def next_page_url
+    return dietary_requirements_url if answer_next_question?
+    return check_your_answers_url if session[:check_answers_seen]
+
+    basic_care_needs_url
+  end
+```
+
+We want anyone to easily change the routing of questions, and the conditional logic that determines which questions a user sees next.
+This is an important but complex part of making the apps more data driven.
+
+## Problem
+
+At the moment all questions are coded as a separate controller. This makes adding new questions or changing the order quite cumbersome.
+Secondly, if we deploy changes to the order then current users could end up completing their journey without answering all questions.
+We would like to solve these by using configuration files (which in the future could be created by a publishing app) to define the flow.  
+
+## Proposal
+
+It is proposed that a YAML file can be used to define the flow through the forms. A single controller would handle the pages that are defined in the YAML file.
+It is proposed that an Interactor pattern is used to handle the validation (beyond the scope of this RFC) and return the next path (assuming validation passes) 
+
+### Simple flows
+Many of the forms have a simple flow which involves checking for invalid values on the form and then progressing to the next page if no validation problems are found.
+(The validation is beyond the scope of this RFC) 
+An example of this would be the `name` controller which checks for validation and then proceeds to the `date_of_birth` controller. This type of progress could simply be defined in YAML as
+```yaml
+name:
+  default_next_question: 
+    question_path: date_of_birth
+``` 
+
+(`question_path` would signify the next path handled by the YAML file and `path` could be used to signify a request to a normal rails route)
+
+### Simple conditionals
+The next simplest definition would be around a different path being taken depending upon an answer selected and stored
+For example, if the user does not live in england we direct them to `not_eligible_england` controller 
+```yaml
+live_in_england:
+  default_next_question: 
+    question_path: nhs_letter
+  conditional_paths:
+  - conditions:
+      live_in_england:
+      - 'No'
+      question_path: not_eligible_england
+```
+
+More complex conditionals could be combined using `and` and `or` logic, similar to [JsonLogic](http://jsonlogic.com/) but is currently
+not proposed (see Custom Routing) and instead is something to consider as we further develop (YAGNI).
+
+### Check answers seen
+A further complexity for the routing is that the user may return to a page/section from the "Check Answers" page. 
+Any implementation must be able to recognise this and return to the check answers page after a section is complete.
+The proposal is that each path can be marked with a flag for `redirect_if_check_answers_seen: true` which will move back to the check answers page as it currently does
```

Comment:
I think there are only a handful of questions that don't route users straight back to the check answers page, so perhaps the exception could be defined in the yaml, and default behaviour assumed for the rest?